File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.py
import sys
input = sys.stdin.readline
from collections import defaultdict

N = int(input())
fruits = list(map(int, input().split()))

left = 0
max_len = 0
counter = defaultdict(int)

for right in range(N):
    counter[fruits[right]] += 1
    
    while len(counter) > 2:
        counter[fruits[left]] -= 1
        if counter[fruits[left]] == 0:
            del counter[fruits[left]]
        left += 1
    
    max_len = max(max_len, right - left + 1)

print(max_len)

File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.js
const input = require('fs').readFileSync(0).toString().trim().split('\n');
const N = Number(input[0]);
const fruits = input[1].split(' ').map(Number);

let left = 0;
let maxLength = 0;
const counter = new Map();

for (let right = 0; right < N; right++) {
    const fruit = fruits[right];
    counter.set(fruit, (counter.get(fruit) || 0) + 1);
    
    while (counter.size > 2) {
        const leftFruit = fruits[left];
        counter.set(leftFruit, counter.get(leftFruit) - 1);
        if (counter.get(leftFruit) === 0) {
            counter.delete(leftFruit)
        }
        left++;
    }
    maxLength = Math.max(maxLength, right - left + 1);
}
console.log(maxLength);